hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/GeneratedValuesProcessor.java

```diff
@@ -187,10 +191,41 @@ private JdbcParameterBindings getJdbcParameterBindings(Object id, SharedSessionC
 		return jdbcParamBindings;
 	}
 
-	private void setEntityAttributes(Object entity, Object[] state, Object[] selectionResults) {
+	private void setEntityAttributes(
+			Object entity,
+			Object[] state,
+			Object[] selectionResults,
+			EventType timing,
+			SharedSessionContractImplementor session) {
 		for ( int i = 0; i < generatedValuesToSelect.size(); i++ ) {
 			final AttributeMapping attribute = generatedValuesToSelect.get( i );
-			final Object generatedValue = selectionResults[i];
+			Object generatedValue = selectionResults[i];
+			if ( attribute.isEmbeddedAttributeMapping() ) {
+				final EmbeddableMappingTypeImpl embeddableMappingType = (EmbeddableMappingTypeImpl) attribute.getMappedType();
+				final Object previousState = state[attribute.getStateArrayPosition()];
+				if ( generatedValue == null && previousState != null ) {
+					generatedValue = embeddableMappingType.getRepresentationStrategy()
+							.getInstantiator()
+							.instantiate(
+									() -> embeddableMappingType.getValues( previousState ),
+									session.getSessionFactory()
+							);
+				}
```

Comment:
Maybe clarify when this can happen? It's not clear why we need to instantiate a new embeddable here.